driver/addlayernorm_driver.hpp

```diff
@@ -0,0 +1,531 @@
+
+    SetTensorNd(inputDesc, in_len, data_type);
+    SetTensorNd(input2Desc, in_len, data_type);
+    SetTensorNd(weightDesc, inner_len, data_type);
+    SetTensorNd(biasDesc, inner_len, data_type);
+    SetTensorNd(outputDesc, in_len, data_type);
+    SetTensorNd(meanDesc, outer_len, data_type);
+    SetTensorNd(rstdDesc, outer_len, data_type);
+
+    eps  = static_cast<double>(inflags.GetValueDouble("eps"));
+    mode = miopenNormMode_t(inflags.GetValueInt("mode"));
+
+    return 0;
+}
+
+template <typename Tgpu, typename Tref>
+int AddLayerNormDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "1", "Run only Forward AddLayerNorm (Default=1)", "int");
+    inflags.AddInputFlag("batchsize", 'n', "100", "Mini-batch size (Default=100)", "int");
+    inflags.AddInputFlag("in_channels", 'c', "3", "Number of Input Channels (Default=3)", "int");
+    inflags.AddInputFlag("in_d", 'D', "0", "Input Depth (Default=0)", "int");
+    inflags.AddInputFlag("in_h", 'H', "32", "Input Height (Default=32)", "int");
+    inflags.AddInputFlag("in_w", 'W', "32", "Input Width (Default=32)", "int");
+
+    inflags.AddInputFlag("eps", 'e', "0.00001", "Alpha (Default=0.00001)", "double");
+    inflags.AddInputFlag("normalized_dim", 'o', "3", "Nomalized Dim (Default=3)", "int");
+    inflags.AddInputFlag(
+        "mode", 'm', "0", "elemwise affine mode (0), weight and bias mode (1) (Default=0)", "int");
+
+    inflags.AddInputFlag("iter", 'i', "10", "Number of Iterations (Default=10)", "int");
+    inflags.AddInputFlag("verify", 'V', "1", "Verify Each Layer (Default=1)", "int");
+    inflags.AddInputFlag("time", 't', "0", "Time Each Layer (Default=0)", "int");
+    inflags.AddInputFlag(
+        "wall", 'w', "0", "Wall-clock Time Each Layer, Requires time == 1 (Default=0)", "int");
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+std::vector<int> AddLayerNormDriver<Tgpu, Tref>::GetInputTensorLengthsFromCmdLine()
+{
+    int in_n = inflags.GetValueInt("batchsize");
+    int in_c = inflags.GetValueInt("in_channels");
+    int in_w = inflags.GetValueInt("in_w");
+    int in_h = inflags.GetValueInt("in_h");
+    int in_d = inflags.GetValueInt("in_d");
+
+    if((in_n != 0) && (in_c != 0) && (in_d != 0) && (in_h != 0) && (in_w != 0))
+    {
+        dim_size = 5;
+        return std::vector<int>({in_n, in_c, in_d, in_h, in_w});
+    }
+    else if((in_n != 0) && (in_c != 0) && (in_h != 0) && (in_w != 0))
+    {
+        dim_size = 4;
+        return std::vector<int>({in_n, in_c, in_h, in_w});
+    }
+    else if((in_n != 0) && (in_c != 0) && (in_w != 0))
+    {
+        dim_size = 3;
+        return std::vector<int>({in_n, in_c, in_w});
+    }
+    else if((in_n != 0) && (in_w != 0))
+    {
+        dim_size = 2;
+        return std::vector<int>({in_n, in_w});
+    }
+    else if(in_n != 0)
+    {
+        return std::vector<int>({in_n});
```

Comment:
I don't see a default value for dim_size, so it is better to set it here. Actually, I think inverting the logic and rewriting these is much more readable. I'm not insisting though:
```suggestion
dim_size = 0;
std::vector<int> dims = {in_n, in_w, in_c, in_h, in_d};
for(int dim : dims)
{
  if(dim == 0)
    break;
  else
    dim_size++;
}

switch(dim_size)
{
  case 1:
    return std::vector<int>({in_n});
  case 2:
    return std::vector<int>({in_n, in_w});
  case 3:
    return std::vector<int>({in_n, in_c, in_w});
  case 4:
    return std::vector<int>({in_n, in_c, in_h, in_w});
  case 5:
    return std::vector<int>({in_n, in_c, in_d, in_h, in_w});
  default:
    std::cout << "Error Input Tensor Lengths\n" << std::endl;   // we actually may want to THROW here
    return std::vector<int>({0});
}
```

Reply:
The method of receiving tensor input has been changed.